File: src/main/java/net/kigawa/util/plugin/stage/StageManager.java
package net.kigawa.util.plugin.stage;

import net.kigawa.util.plugin.KigawaPlugin;
import net.kigawa.util.plugin.recorder.EqualsRecorderData;
import org.bukkit.command.CommandSender;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class StageManager {
    List<StageData> allStage=new ArrayList<>();
    List<StageData> canUse=new ArrayList<>();
    List<StageData> notUse=new ArrayList<>();
    KigawaPlugin plugin;
    public StageManager(KigawaPlugin kigawaPlugin){
        plugin=kigawaPlugin;

        File folder=new File( plugin.getDataFolder(),"stage");
        folder.mkdir();
        String[] files=folder.list();
        for (int i=0;i<files.length;i++){
            File file=new File(folder,files[i]);
            plugin.logger(files[i]);
            StageData data=(StageData) plugin.getRecorder().load(StageData.class,"stage",files[i].substring(0, files[i].length()-4));
            canUse.add(data);
            allStage.add(data);
        }
    }
    public void setStage(String name, CommandSender sender){
        //check can use this name
        if (plugin.getStageManager().getStage(name)!=null){
            //create Stage
            StageData stageData=new StageData();
            stageData.setName(name);
            //put in list
            canUse.add(stageData);
            allStage.add(stageData);
            //save
            plugin.getRecorder().save(stageData);
        }else {
            sender.sendMessage("this name can't use");
        }
    }
    public StageData getStage(String name){
        StageData stageData=null;
        if (allStage.contains(new EqualsRecorderData(name))){
            stageData= allStage.get(allStage.indexOf(new EqualsRecorderData(name)));
        }
        return stageData;
    }
    public StageData getRandomStage(){
        StageData stageData=null;
         if (canUse.size()>0){
             Random random=new Random();
             int randomNumber= random.nextInt(canUse.size());
             stageData=canUse.get(randomNumber);
             notUse.add(stageData);
             canUse.remove(randomNumber);
         }
         return stageData;
    }
}

File: src/main/java/net/kigawa/onigoplugin/util/plugin/game/stage/StageManager.kt
package net.kigawa.onigoplugin.util.plugin.game.stage

import net.kigawa.kutil.kutil.list.contains
import net.kigawa.kutil.unitapi.annotation.Kunit
import net.kigawa.onigoplugin.OnigoPlugin
import net.kigawa.onigoplugin.util.plugin.all.recorder.Recorder
import org.bukkit.command.CommandSender
import java.io.File
import java.util.*

@Kunit
class StageManager(var plugin: OnigoPlugin, private val recorder: Recorder) {
  val allStage: MutableList<StageData> = ArrayList()
  private var canUse: MutableList<StageData> = ArrayList()
  private var notUse: MutableList<StageData> = ArrayList()

  init {
    val folder = File(plugin.dataFolder, "stage")
    folder.mkdir()
    folder.list()?.forEach {
      plugin.logger(it)
      val data = recorder.load(StageData::class.java, "stage", it.substring(0, it.length - 4)) as StageData
      canUse.add(data)
      allStage.add(data)
    }
  }

  fun save(stageData: StageData) {
    recorder.save(stageData, "stage")
  }

  fun setStage(name: String, sender: CommandSender) {
    //check can use this name
    if (getStage(name) == null) {
      //create Stage
      val stageData = StageData(name)
      //put in list
      canUse.add(stageData)
      allStage.add(stageData)
      //save
      save(stageData)
    } else {
      sender.sendMessage("this name can't use")
    }
  }

  fun getStage(name: String?): StageData? {
    var stageData: StageData? = null
    if (name != null) {
      stageData = allStage.firstOrNull { it.name == name }
    }
    if (name == null) {
      stageData = randomStage
    }
    return stageData
  }

  val randomStage: StageData?
    get() {
      var stageData: StageData? = null
      if (canUse.size > 0) {
        val random = Random()
        val randomNumber = random.nextInt(canUse.size)
        stageData = canUse[randomNumber]
        notUse.add(stageData)
        canUse.removeAt(randomNumber)
      }
      return stageData
    }

  fun returnStage(stageData: StageData) {
    if (notUse.contains { it.name == stageData.name }) {
      canUse.add(stageData)
      notUse.removeIf { it.name == stageData.name }
    }
  }
}
